File: src/main/java/com/sulabh/sulabh_backend/service/ComplaintService.java
package com.sulabh.sulabh_backend.service;

import com.sulabh.sulabh_backend.dto.ComplaintRequest;
import com.sulabh.sulabh_backend.dto.ComplaintResponse;
import com.sulabh.sulabh_backend.dto.ComplaintUpdateRequest;
import com.sulabh.sulabh_backend.dto.ComplaintUpdateResponse;
import com.sulabh.sulabh_backend.entity.Complaint;
import com.sulabh.sulabh_backend.entity.ComplaintUpdate;
import com.sulabh.sulabh_backend.entity.User;
import com.sulabh.sulabh_backend.repository.ComplaintRepository;
import com.sulabh.sulabh_backend.repository.ComplaintUpdateRepository;
import com.sulabh.sulabh_backend.util.ResponseMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class ComplaintService {
    
    private final ComplaintRepository complaintRepository;
    private final ComplaintUpdateRepository complaintUpdateRepository;
    private final ResponseMapper responseMapper;
    
    public ComplaintResponse submitComplaint(ComplaintRequest request, User user) {
        Complaint complaint = Complaint.builder()
                .user(user)
                .category(request.getCategory())
                .subject(request.getSubject())
                .description(request.getDescription())
                .location(request.getLocation())
                .priority(request.getPriority())
                .status(Complaint.ComplaintStatus.PENDING)
                .attachments(request.getAttachments())
                .submittedAt(new Date())
                .updatedAt(new Date())
                .build();
        
        complaint = complaintRepository.save(complaint);
        
        // Create initial update
        ComplaintUpdate initialUpdate = ComplaintUpdate.builder()
                .complaint(complaint)
                .message("Complaint submitted successfully")
                .status(Complaint.ComplaintStatus.PENDING)
                .updatedBy(user.getEmail())
                .updatedAt(new Date())
                .build();
        
        complaintUpdateRepository.save(initialUpdate);
        
        return responseMapper.toComplaintResponse(complaint);
    }
    
    public ComplaintResponse getComplaintById(Long id) {
        Complaint complaint = complaintRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Complaint not found with id: " + id));
        return responseMapper.toComplaintResponse(complaint);
    }
    
    public List<ComplaintResponse> getComplaintsByUser(User user) {
        List<Complaint> complaints = complaintRepository.findByUserOrderBySubmittedAtDesc(user);
        return complaints.stream()
                .map(responseMapper::toComplaintResponse)
                .collect(Collectors.toList());
    }
    
    public Page<ComplaintResponse> getComplaintsByUser(User user, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        Page<Complaint> complaints = complaintRepository.findByUserOrderBySubmittedAtDesc(user, pageable);
        return complaints.map(responseMapper::toComplaintResponse);
    }
    
    public List<ComplaintResponse> getAllComplaints() {
        List<Complaint> complaints = complaintRepository.findAll();
        return complaints.stream()
                .map(responseMapper::toComplaintResponse)
                .collect(Collectors.toList());
    }
    
    public List<ComplaintResponse> getComplaintsByStatus(Complaint.ComplaintStatus status) {
        List<Complaint> complaints = complaintRepository.findByStatusOrderBySubmittedAtDesc(status);
        return complaints.stream()
                .map(responseMapper::toComplaintResponse)
                .collect(Collectors.toList());
    }
    
    public ComplaintUpdateResponse addComplaintUpdate(Long complaintId, ComplaintUpdateRequest request, String updatedBy) {
        Complaint complaint = complaintRepository.findById(complaintId)
                .orElseThrow(() -> new RuntimeException("Complaint not found with id: " + complaintId));
        
        ComplaintUpdate update = ComplaintUpdate.builder()
                .complaint(complaint)
                .message(request.getMessage())
                .status(request.getStatus())
                .updatedBy(updatedBy)
                .attachments(request.getAttachments())
                .updatedAt(new Date())
                .build();
        
        update = complaintUpdateRepository.save(update);
        
        // Update complaint status if provided
        if (request.getStatus() != null) {
            complaint.setStatus(request.getStatus());
            complaint.setUpdatedAt(new Date());
            if (request.getStatus() == Complaint.ComplaintStatus.RESOLVED) {
                complaint.setResolvedAt(new Date());
            }
            complaintRepository.save(complaint);
        }
        
        return responseMapper.toComplaintUpdateResponse(update);
    }
    
    public List<ComplaintUpdateResponse> getComplaintUpdates(Long complaintId) {
        List<ComplaintUpdate> updates = complaintUpdateRepository.findByComplaintIdOrderByUpdatedAtDesc(complaintId);
        return updates.stream()
                .map(responseMapper::toComplaintUpdateResponse)
                .collect(Collectors.toList());
    }
    
    public ComplaintResponse updateComplaintStatus(Long complaintId, Complaint.ComplaintStatus status, String updatedBy) {
        Complaint complaint = complaintRepository.findById(complaintId)
                .orElseThrow(() -> new RuntimeException("Complaint not found with id: " + complaintId));
        
        complaint.setStatus(status);
        complaint.setUpdatedAt(new Date());
        
        if (status == Complaint.ComplaintStatus.RESOLVED) {
            complaint.setResolvedAt(new Date());
        }
        
        complaint = complaintRepository.save(complaint);
        
        // Add status update
        ComplaintUpdate update = ComplaintUpdate.builder()
                .complaint(complaint)
                .message("Status updated to " + status.name())
                .status(status)
                .updatedBy(updatedBy)
                .updatedAt(new Date())
                .build();
        
        complaintUpdateRepository.save(update);
        
        return responseMapper.toComplaintResponse(complaint);
    }
    
    public void deleteComplaint(Long complaintId, User user) {
        Complaint complaint = complaintRepository.findById(complaintId)
                .orElseThrow(() -> new RuntimeException("Complaint not found with id: " + complaintId));
        
        // Only allow user to delete their own complaints and only if it's pending
        if (!complaint.getUser().getId().equals(user.getId())) {
            throw new RuntimeException("You can only delete your own complaints");
        }
        
        if (complaint.getStatus() != Complaint.ComplaintStatus.PENDING) {
            throw new RuntimeException("You can only delete pending complaints");
        }
        
        complaintRepository.delete(complaint);
    }
    
    // Statistics methods
    public Long getTotalComplaintsCount() {
        return complaintRepository.count();
    }
    
    public Long getComplaintsCountByStatus(Complaint.ComplaintStatus status) {
        return complaintRepository.countByStatus(status);
    }
    
    public Long getUserComplaintsCount(User user) {
        return complaintRepository.countByUser(user);
    }
    
    public Long getUserComplaintsCountByStatus(User user, Complaint.ComplaintStatus status) {
        return complaintRepository.countByUserAndStatus(user, status);
    }
}

File: src/services/complaintService.ts
import api from './api'

export interface ComplaintRequest {
  category: string
  subject: string
  description: string
  location: string
  priority?: string
  attachments?: string[]
}

export interface ComplaintResponse {
  id: number
  category: string
  subject: string
  description: string
  location: string
  priority: string
  status: string
  assignedTo?: string
  attachments: string[]
  submittedAt: string
  updatedAt: string
  resolvedAt?: string
  user: {
    id: number
    email: string
    username: string
    fullName: string
  }
  updates: ComplaintUpdateResponse[]
}

export interface ComplaintUpdateResponse {
  id: number
  message: string
  status?: string
  updatedBy: string
  attachments: string[]
  updatedAt: string
}

export interface ComplaintUpdateRequest {
  message: string
  status?: string
  attachments?: string[]
}

class ComplaintService {
  
  // Submit a new complaint
  async submitComplaint(complaintData: ComplaintRequest): Promise<ComplaintResponse> {
    try {
      const response = await api.post<ComplaintResponse>('/api/complaints', complaintData)
      return response.data
    } catch (error: any) {
      console.error('Submit complaint error:', error)
      if (error.response?.data?.message) {
        throw new Error(error.response.data.message)
      } else {
        throw new Error('Failed to submit complaint. Please try again.')
      }
    }
  }

  // Get complaint by ID
  async getComplaint(id: string): Promise<ComplaintResponse> {
    try {
      const response = await api.get<ComplaintResponse>(`/api/complaints/${id}`)
      return response.data
    } catch (error: any) {
      console.error('Get complaint error:', error)
      if (error.response?.status === 404) {
        throw new Error('Complaint not found')
      } else {
        throw new Error('Failed to fetch complaint. Please try again.')
      }
    }
  }

  // Get user's complaints
  async getMyComplaints(): Promise<ComplaintResponse[]> {
    try {
      const response = await api.get<ComplaintResponse[]>('/api/complaints/my')
      return response.data
    } catch (error: any) {
      console.error('Get my complaints error:', error)
      throw new Error('Failed to fetch complaints. Please try again.')
    }
  }

  // Get user's complaints with pagination
  async getMyComplaintsPaginated(page: number = 0, size: number = 10): Promise<{
    content: ComplaintResponse[]
    totalElements: number
    totalPages: number
    size: number
    number: number
  }> {
    try {
      const response = await api.get(`/api/complaints/my/paginated?page=${page}&size=${size}`)
      return response.data
    } catch (error: any) {
      console.error('Get paginated complaints error:', error)
      throw new Error('Failed to fetch complaints. Please try again.')
    }
  }

  // Get all complaints (admin)
  async getAllComplaints(): Promise<ComplaintResponse[]> {
    try {
      const response = await api.get<ComplaintResponse[]>('/api/complaints')
      return response.data
    } catch (error: any) {
      console.error('Get all complaints error:', error)
      throw new Error('Failed to fetch complaints. Please try again.')
    }
  }

  // Get complaints by status
  async getComplaintsByStatus(status: string): Promise<ComplaintResponse[]> {
    try {
      const response = await api.get<ComplaintResponse[]>(`/api/complaints/status/${status}`)
      return response.data
    } catch (error: any) {
      console.error('Get complaints by status error:', error)
      throw new Error('Failed to fetch complaints. Please try again.')
    }
  }

  // Add update to complaint
  async addComplaintUpdate(complaintId: string, updateData: ComplaintUpdateRequest): Promise<ComplaintUpdateResponse> {
    try {
      const response = await api.post<ComplaintUpdateResponse>(`/api/complaints/${complaintId}/updates`, updateData)
      return response.data
    } catch (error: any) {
      console.error('Add complaint update error:', error)
      if (error.response?.data?.message) {
        throw new Error(error.response.data.message)
      } else {
        throw new Error('Failed to add update. Please try again.')
      }
    }
  }

  // Get complaint updates
  async getComplaintUpdates(complaintId: string): Promise<ComplaintUpdateResponse[]> {
    try {
      const response = await api.get<ComplaintUpdateResponse[]>(`/api/complaints/${complaintId}/updates`)
      return response.data
    } catch (error: any) {
      console.error('Get complaint updates error:', error)
      throw new Error('Failed to fetch updates. Please try again.')
    }
  }

  // Update complaint status
  async updateComplaintStatus(complaintId: string, status: string): Promise<ComplaintResponse> {
    try {
      const response = await api.put<ComplaintResponse>(`/api/complaints/${complaintId}/status`, { status })
      return response.data
    } catch (error: any) {
      console.error('Update complaint status error:', error)
      if (error.response?.data?.message) {
        throw new Error(error.response.data.message)
      } else {
        throw new Error('Failed to update status. Please try again.')
      }
    }
  }

  // Delete complaint
  async deleteComplaint(complaintId: string): Promise<void> {
    try {
      await api.delete(`/api/complaints/${complaintId}`)
    } catch (error: any) {
      console.error('Delete complaint error:', error)
      if (error.response?.data?.message) {
        throw new Error(error.response.data.message)
      } else {
        throw new Error('Failed to delete complaint. Please try again.')
      }
    }
  }

  // Get complaint statistics
  async getComplaintStats(): Promise<{
    total: number
    pending: number
    inProgress: number
    resolved: number
    closed: number
  }> {
    try {
      const response = await api.get('/api/complaints/stats')
      return response.data
    } catch (error: any) {
      console.error('Get complaint stats error:', error)
      throw new Error('Failed to fetch statistics. Please try again.')
    }
  }

  // Get admin complaint statistics
  async getAdminComplaintStats(): Promise<{
    total: number
    pending: number
    inProgress: number
    resolved: number
    closed: number
  }> {
    try {
      const response = await api.get('/api/complaints/admin/stats')
      return response.data
    } catch (error: any) {
      console.error('Get admin complaint stats error:', error)
      throw new Error('Failed to fetch statistics. Please try again.')
    }
  }
}

// Export singleton instance
export default new ComplaintService()